File: pkg/adapters/playwright/python/src/traffic2openapi_playwright/capture.py
"""
Playwright traffic capture module.
"""

import json
import re
import time
from dataclasses import dataclass, field
from datetime import datetime
from pathlib import Path
from typing import Callable, Optional, Union
from urllib.parse import parse_qs, urlparse

from playwright.sync_api import BrowserContext, Page, Request as PWRequest, Response as PWResponse
from playwright.async_api import BrowserContext as AsyncBrowserContext

from .types import IRRecord, Request, Response, RequestMethod
from .writer import NDJSONWriter, GzipNDJSONWriter


# Default headers to exclude (security-sensitive)
DEFAULT_EXCLUDE_HEADERS = frozenset([
    "authorization",
    "cookie",
    "set-cookie",
    "x-api-key",
    "x-auth-token",
    "x-csrf-token",
    "proxy-authorization",
])


@dataclass
class CaptureOptions:
    """Configuration options for traffic capture."""

    # Output file path
    output: Union[str, Path]

    # Filter options
    filter_hosts: list[str] = field(default_factory=list)
    filter_methods: list[str] = field(default_factory=list)
    exclude_paths: list[str] = field(default_factory=list)
    exclude_path_patterns: list[re.Pattern] = field(default_factory=list)

    # Header filtering
    exclude_headers: set[str] = field(default_factory=lambda: set(DEFAULT_EXCLUDE_HEADERS))
    include_headers: bool = True

    # Body capture
    capture_request_body: bool = True
    capture_response_body: bool = True
    max_body_size: int = 1024 * 1024  # 1MB

    # Content type filtering for body capture
    capture_content_types: list[str] = field(default_factory=lambda: [
        "application/json",
        "application/xml",
        "text/xml",
        "text/plain",
        "text/html",
    ])

    # Compression
    gzip: bool = False
    compression_level: int = 9

    # Error handling
    on_error: Optional[Callable[[Exception], None]] = None


class PlaywrightCapture:
    """
    Captures HTTP traffic from Playwright and writes IR records.

    Usage (sync):
        with sync_playwright() as p:
            browser = p.chromium.launch()
            context = browser.new_context()

            capture = PlaywrightCapture("traffic.ndjson")
            capture.attach(context)

            page = context.new_page()
            page.goto("https://api.example.com")

            capture.close()
            browser.close()

    Usage (async):
        async with async_playwright() as p:
            browser = await p.chromium.launch()
            context = await browser.new_context()

            capture = PlaywrightCapture("traffic.ndjson")
            await capture.attach_async(context)

            page = await context.new_page()
            await page.goto("https://api.example.com")

            capture.close()
            await browser.close()
    """

    def __init__(
        self,
        output: Union[str, Path, CaptureOptions],
        **kwargs,
    ):
        """
        Initialize traffic capture.

        Args:
            output: Output file path or CaptureOptions.
            **kwargs: Additional options (if output is a path).
        """
        if isinstance(output, CaptureOptions):
            self.options = output
        else:
            self.options = CaptureOptions(output=output, **kwargs)

        # Create writer
        if self.options.gzip or str(self.options.output).endswith(".gz"):
            self._writer = GzipNDJSONWriter(
                self.options.output,
                compression_level=self.options.compression_level,
            )
        else:
            self._writer = NDJSONWriter(self.options.output)

        # Track pending requests (for timing)
        self._pending_requests: dict[str, tuple[PWRequest, float]] = {}

    def attach(self, context: BrowserContext) -> None:
        """
        Attach to a Playwright browser context (sync).

        Args:
            context: Playwright BrowserContext to capture traffic from.
        """
        context.on("request", self._on_request)
        context.on("response", self._on_response)

    async def attach_async(self, context: AsyncBrowserContext) -> None:
        """
        Attach to a Playwright browser context (async).

        Args:
            context: Playwright async BrowserContext to capture traffic from.
        """
        context.on("request", self._on_request)
        context.on("response", self._on_response_async)

    def _on_request(self, request: PWRequest) -> None:
        """Handle request event."""
        # Store request with timestamp for duration calculation
        self._pending_requests[request.url] = (request, time.time())

    def _on_response(self, response: PWResponse) -> None:
        """Handle response event (sync)."""
        try:
            record = self._create_record(response)
            if record:
                self._writer.write(record)
        except Exception as e:
            if self.options.on_error:
                self.options.on_error(e)

    async def _on_response_async(self, response: PWResponse) -> None:
        """Handle response event (async)."""
        try:
            record = await self._create_record_async(response)
            if record:
                self._writer.write(record)
        except Exception as e:
            if self.options.on_error:
                self.options.on_error(e)

    def _should_capture(self, request: PWRequest) -> bool:
        """Check if request should be captured based on filters."""
        url = urlparse(request.url)

        # Host filter
        if self.options.filter_hosts:
            if url.hostname not in self.options.filter_hosts:
                return False

        # Method filter
        if self.options.filter_methods:
            if request.method.upper() not in [m.upper() for m in self.options.filter_methods]:
                return False

        # Path exclusion
        if self.options.exclude_paths:
            if url.path in self.options.exclude_paths:
                return False

        # Path pattern exclusion
        for pattern in self.options.exclude_path_patterns:
            if pattern.match(url.path):
                return False

        return True

    def _filter_headers(self, headers: dict[str, str]) -> dict[str, str]:
        """Filter out excluded headers."""
        if not self.options.include_headers:
            return {}
        return {
            k.lower(): v
            for k, v in headers.items()
            if k.lower() not in self.options.exclude_headers
        }

    def _should_capture_body(self, content_type: Optional[str]) -> bool:
        """Check if body should be captured based on content type."""
        if not content_type:
            return False
        # Check if content type starts with any allowed type
        for allowed in self.options.capture_content_types:
            if content_type.startswith(allowed):
                return True
        return False

    def _parse_body(self, body: bytes, content_type: Optional[str]) -> Optional[any]:
        """Parse body based on content type."""
        if not body or len(body) > self.options.max_body_size:
            return None

        try:
            if content_type and "json" in content_type:
                return json.loads(body.decode("utf-8"))
            else:
                return body.decode("utf-8")
        except (json.JSONDecodeError, UnicodeDecodeError):
            return None

    def _create_record(self, response: PWResponse) -> Optional[IRRecord]:
        """Create IR record from Playwright response (sync)."""
        request = response.request

        if not self._should_capture(request):
            return None

        # Get timing info
        start_time = None
        if request.url in self._pending_requests:
            _, start_time = self._pending_requests.pop(request.url)

        url = urlparse(request.url)

        # Parse query parameters
        query_params = {}
        if url.query:
            parsed = parse_qs(url.query)
            query_params = {k: v[0] if len(v) == 1 else v for k, v in parsed.items()}

        # Get request body
        request_body = None
        if self.options.capture_request_body:
            try:
                post_data = request.post_data
                if post_data:
                    content_type = request.headers.get("content-type", "")
                    if self._should_capture_body(content_type):
                        if isinstance(post_data, str):
                            post_data = post_data.encode("utf-8")
                        request_body = self._parse_body(post_data, content_type)
            except Exception:
                pass

        # Get response body
        response_body = None
        if self.options.capture_response_body:
            try:
                content_type = response.headers.get("content-type", "")
                if self._should_capture_body(content_type):
                    body = response.body()
                    response_body = self._parse_body(body, content_type)
            except Exception:
                pass

        # Calculate duration
        duration_ms = None
        if start_time:
            duration_ms = (time.time() - start_time) * 1000

        return IRRecord(
            request=Request(
                method=RequestMethod(request.method.upper()),
                path=url.path or "/",
                scheme=url.scheme,
                host=url.hostname,
                query=query_params if query_params else None,
                headers=self._filter_headers(request.headers) or None,
                content_type=request.headers.get("content-type"),
                body=request_body,
            ),
            response=Response(
                status=response.status,
                headers=self._filter_headers(response.headers) or None,
                content_type=response.headers.get("content-type"),
                body=response_body,
            ),
            duration_ms=duration_ms,
        )

    async def _create_record_async(self, response: PWResponse) -> Optional[IRRecord]:
        """Create IR record from Playwright response (async)."""
        request = response.request

        if not self._should_capture(request):
            return None

        # Get timing info
        start_time = None
        if request.url in self._pending_requests:
            _, start_time = self._pending_requests.pop(request.url)

        url = urlparse(request.url)

        # Parse query parameters
        query_params = {}
        if url.query:
            parsed = parse_qs(url.query)
            query_params = {k: v[0] if len(v) == 1 else v for k, v in parsed.items()}

        # Get request body
        request_body = None
        if self.options.capture_request_body:
            try:
                post_data = request.post_data
                if post_data:
                    content_type = request.headers.get("content-type", "")
                    if self._should_capture_body(content_type):
                        if isinstance(post_data, str):
                            post_data = post_data.encode("utf-8")
                        request_body = self._parse_body(post_data, content_type)
            except Exception:
                pass

        # Get response body (async)
        response_body = None
        if self.options.capture_response_body:
            try:
                content_type = response.headers.get("content-type", "")
                if self._should_capture_body(content_type):
                    body = await response.body()
                    response_body = self._parse_body(body, content_type)
            except Exception:
                pass

        # Calculate duration
        duration_ms = None
        if start_time:
            duration_ms = (time.time() - start_time) * 1000

        return IRRecord(
            request=Request(
                method=RequestMethod(request.method.upper()),
                path=url.path or "/",
                scheme=url.scheme,
                host=url.hostname,
                query=query_params if query_params else None,
                headers=self._filter_headers(request.headers) or None,
                content_type=request.headers.get("content-type"),
                body=request_body,
            ),
            response=Response(
                status=response.status,
                headers=self._filter_headers(response.headers) or None,
                content_type=response.headers.get("content-type"),
                body=response_body,
            ),
            duration_ms=duration_ms,
        )

    def flush(self) -> None:
        """Flush any buffered records."""
        self._writer.flush()

    def close(self) -> None:
        """Close the capture and writer."""
        self._writer.close()

    @property
    def count(self) -> int:
        """Number of records captured."""
        return self._writer.count

    def __enter__(self) -> "PlaywrightCapture":
        return self

    def __exit__(self, exc_type, exc_val, exc_tb) -> None:
        self.close()

File: pkg/adapters/playwright/typescript/src/capture.ts
/**
 * Playwright traffic capture module.
 */

import type { BrowserContext, Request as PWRequest, Response as PWResponse } from 'playwright';

import { IRRecord, Request, Response, RequestMethod, createIRRecord } from './types.js';
import { NDJSONWriter, GzipNDJSONWriter } from './writer.js';

/**
 * Default headers to exclude (security-sensitive).
 */
export const DEFAULT_EXCLUDE_HEADERS = new Set([
  'authorization',
  'cookie',
  'set-cookie',
  'x-api-key',
  'x-auth-token',
  'x-csrf-token',
  'proxy-authorization',
]);

/**
 * Configuration options for traffic capture.
 */
export interface CaptureOptions {
  /** Output file path. */
  output: string;

  /** Only capture requests to these hosts. */
  filterHosts?: string[];

  /** Only capture these HTTP methods. */
  filterMethods?: RequestMethod[];

  /** Skip requests matching these paths. */
  excludePaths?: string[];

  /** Skip requests matching these path patterns. */
  excludePathPatterns?: RegExp[];

  /** Headers to exclude from capture. */
  excludeHeaders?: Set<string>;

  /** Include headers in output (default: true). */
  includeHeaders?: boolean;

  /** Capture request bodies (default: true). */
  captureRequestBody?: boolean;

  /** Capture response bodies (default: true). */
  captureResponseBody?: boolean;

  /** Max body size in bytes (default: 1MB). */
  maxBodySize?: number;

  /** Content types to capture bodies for. */
  captureContentTypes?: string[];

  /** Use gzip compression (default: false). */
  gzip?: boolean;

  /** Gzip compression level 1-9 (default: 9). */
  compressionLevel?: number;

  /** Error handler callback. */
  onError?: (error: Error) => void;
}

/**
 * Default capture options.
 */
const DEFAULT_OPTIONS: Partial<CaptureOptions> = {
  excludeHeaders: DEFAULT_EXCLUDE_HEADERS,
  includeHeaders: true,
  captureRequestBody: true,
  captureResponseBody: true,
  maxBodySize: 1024 * 1024, // 1MB
  captureContentTypes: [
    'application/json',
    'application/xml',
    'text/xml',
    'text/plain',
    'text/html',
  ],
  gzip: false,
  compressionLevel: 9,
};

/**
 * Captures HTTP traffic from Playwright and writes IR records.
 *
 * @example
 * ```typescript
 * import { chromium } from 'playwright';
 * import { PlaywrightCapture } from 'traffic2openapi-playwright';
 *
 * const browser = await chromium.launch();
 * const context = await browser.newContext();
 *
 * const capture = new PlaywrightCapture({ output: 'traffic.ndjson' });
 * capture.attach(context);
 *
 * const page = await context.newPage();
 * await page.goto('https://api.example.com');
 *
 * await capture.close();
 * await browser.close();
 * ```
 */
export class PlaywrightCapture {
  private options: Required<CaptureOptions>;
  private writer: NDJSONWriter | GzipNDJSONWriter;
  private pendingRequests = new Map<string, { request: PWRequest; startTime: number }>();

  /**
   * Create a new traffic capture instance.
   * @param options - Capture options or output file path.
   */
  constructor(options: CaptureOptions | string) {
    if (typeof options === 'string') {
      options = { output: options };
    }

    this.options = { ...DEFAULT_OPTIONS, ...options } as Required<CaptureOptions>;

    // Create writer
    if (this.options.gzip || this.options.output.endsWith('.gz')) {
      this.writer = new GzipNDJSONWriter(this.options.output, this.options.compressionLevel);
    } else {
      this.writer = new NDJSONWriter(this.options.output);
    }
  }

  /**
   * Attach to a Playwright browser context.
   * @param context - Playwright BrowserContext to capture traffic from.
   */
  attach(context: BrowserContext): void {
    context.on('request', (request) => this.onRequest(request));
    context.on('response', (response) => this.onResponse(response));
  }

  private onRequest(request: PWRequest): void {
    // Store request with timestamp for duration calculation
    this.pendingRequests.set(request.url(), {
      request,
      startTime: Date.now(),
    });
  }

  private async onResponse(response: PWResponse): Promise<void> {
    try {
      const record = await this.createRecord(response);
      if (record) {
        this.writer.write(record);
      }
    } catch (error) {
      if (this.options.onError) {
        this.options.onError(error as Error);
      }
    }
  }

  private shouldCapture(request: PWRequest): boolean {
    const url = new URL(request.url());

    // Host filter
    if (this.options.filterHosts?.length) {
      if (!this.options.filterHosts.includes(url.hostname)) {
        return false;
      }
    }

    // Method filter
    if (this.options.filterMethods?.length) {
      if (!this.options.filterMethods.includes(request.method() as RequestMethod)) {
        return false;
      }
    }

    // Path exclusion
    if (this.options.excludePaths?.length) {
      if (this.options.excludePaths.includes(url.pathname)) {
        return false;
      }
    }

    // Path pattern exclusion
    if (this.options.excludePathPatterns?.length) {
      for (const pattern of this.options.excludePathPatterns) {
        if (pattern.test(url.pathname)) {
          return false;
        }
      }
    }

    return true;
  }

  private filterHeaders(headers: Record<string, string>): Record<string, string> | undefined {
    if (!this.options.includeHeaders) {
      return undefined;
    }

    const filtered: Record<string, string> = {};
    for (const [key, value] of Object.entries(headers)) {
      if (!this.options.excludeHeaders.has(key.toLowerCase())) {
        filtered[key.toLowerCase()] = value;
      }
    }
    return Object.keys(filtered).length > 0 ? filtered : undefined;
  }

  private shouldCaptureBody(contentType?: string): boolean {
    if (!contentType) return false;
    return this.options.captureContentTypes.some((type) => contentType.startsWith(type));
  }

  private parseBody(body: Buffer, contentType?: string): unknown {
    if (!body || body.length > this.options.maxBodySize) {
      return undefined;
    }

    try {
      const text = body.toString('utf-8');
      if (contentType?.includes('json')) {
        return JSON.parse(text);
      }
      return text;
    } catch {
      return undefined;
    }
  }

  private async createRecord(response: PWResponse): Promise<IRRecord | null> {
    const pwRequest = response.request();

    if (!this.shouldCapture(pwRequest)) {
      return null;
    }

    // Get timing info
    const pending = this.pendingRequests.get(pwRequest.url());
    if (pending) {
      this.pendingRequests.delete(pwRequest.url());
    }

    const url = new URL(pwRequest.url());

    // Parse query parameters
    const query: Record<string, string | string[]> = {};
    url.searchParams.forEach((value, key) => {
      const existing = query[key];
      if (existing) {
        if (Array.isArray(existing)) {
          existing.push(value);
        } else {
          query[key] = [existing, value];
        }
      } else {
        query[key] = value;
      }
    });

    // Get request body
    let requestBody: unknown;
    if (this.options.captureRequestBody) {
      try {
        const postData = pwRequest.postData();
        if (postData) {
          const contentType = pwRequest.headers()['content-type'];
          if (this.shouldCaptureBody(contentType)) {
            requestBody = this.parseBody(Buffer.from(postData), contentType);
          }
        }
      } catch {
        // Ignore
      }
    }

    // Get response body
    let responseBody: unknown;
    if (this.options.captureResponseBody) {
      try {
        const contentType = response.headers()['content-type'];
        if (this.shouldCaptureBody(contentType)) {
          const body = await response.body();
          responseBody = this.parseBody(body, contentType);
        }
      } catch {
        // Ignore
      }
    }

    // Calculate duration
    const durationMs = pending ? Date.now() - pending.startTime : undefined;

    const request: Request = {
      method: pwRequest.method() as RequestMethod,
      path: url.pathname || '/',
      scheme: url.protocol.replace(':', ''),
      host: url.hostname,
      query: Object.keys(query).length > 0 ? query : undefined,
      headers: this.filterHeaders(pwRequest.headers()),
      contentType: pwRequest.headers()['content-type'],
      body: requestBody,
    };

    const resp: Response = {
      status: response.status(),
      headers: this.filterHeaders(response.headers()),
      contentType: response.headers()['content-type'],
      body: responseBody,
    };

    return createIRRecord(request, resp, { durationMs });
  }

  /**
   * Flush any buffered records.
   */
  async flush(): Promise<void> {
    await this.writer.flush();
  }

  /**
   * Close the capture and writer.
   */
  async close(): Promise<void> {
    await this.writer.close();
  }

  /**
   * Number of records captured.
   */
  get count(): number {
    return this.writer.count;
  }
}
